### ISSUE TITLE
Rethread crash during huge network transfer

### ISSUE THREAD

**Issue by nighelles** [rethinkdb/rethinkdb#5353]
error: Error in ./src/concurrency/coro_pool.hpp at line 57:
error: Assertion failed: [active_worker_count == 0]
error: Backtrace:
error: Wed Feb  3 11:42:14 2016

       1 [0x22c27cb]: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:108
       2 [0x22be45b]: backtrace_t::backtrace_t() at backtrace.cc:210
       3 [0x22bef45]: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:290
       4 [0x22be0e7]: format_backtrace(bool) at backtrace.cc:205
       5 [0x22c68aa]: report_fatal_error(char const*, int, char const*, ...) at errors.cc:85
       6 [0x1637259]: coro_pool_t<linux_tcp_conn_t::write_queue_op_t*>::rethread(threadnum_t) at coro_pool.hpp:57
       7 [0x16333ed]: linux_tcp_conn_t::rethread(threadnum_t) at network.cc:595
       8 [0x1c9a995]: tcp_conn_stream_t::rethread(threadnum_t) at tcp_conn_stream.cc:85
       9 [0x1c9afc4]: rethread_tcp_conn_stream_t::~rethread_tcp_conn_stream_t() at tcp_conn_stream.cc:175
       10 [0x21d1dce]: connectivity_cluster_t::run_t::handle(keepalive_tcp_conn_stream_t*, boost::optional<peer_id_t>, boost::optional<peer_address_t>, auto_drainer_t::lock_t, bool*) at cluster.cc:1240
       11 [0x21d2459]: connectivity_cluster_t::run_t::connect_to_peer(peer_address_t const*, int, boost::optional<peer_id_t>, auto_drainer_t::lock_t, bool*, co_semaphore_t*) at cluster.cc:340
       12 [0x21e32f4]: std::_Mem_fn<void (connectivity_cluster_t::run_t::*)(peer_address_t const*, int, boost::optional<peer_id_t>, auto_drainer_t::lock_t, bool*, co_semaphore_t*)>::operator()(connectivity_cluster_t::run_t*, peer_address_t const*, int, boost::optional<peer_id_t>, auto_drainer_t::lock_t, bool*, co_semaphore_t*) const at functional:550
       13 [0x21e312b]: _ZNKSt5_BindIFSt7_Mem_fnIMN22connectivity_cluster_t5run_tEFvPK14peer_address_tiN5boost8optionalI9peer_id_tEEN14auto_drainer_t6lock_tEPbP14co_semaphore_tEEPS2_S5_St12_PlaceholderILi1EES9_SB_SC_P18static_semaphore_tEE8__call_cIvJRlEJLi0ELi1ELi2ELi3ELi4ELi5ELi6EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0x1db at functional:1154
       14 [0x21e2f3d]: _ZNKSt5_BindIFSt7_Mem_fnIMN22connectivity_cluster_t5run_tEFvPK14peer_address_tiN5boost8optionalI9peer_id_tEEN14auto_drainer_t6lock_tEPbP14co_semaphore_tEEPS2_S5_St12_PlaceholderILi1EES9_SB_SC_P18static_semaphore_tEEclIJRlEvEET0_DpOT_+0x3d at functional:1217
       15 [0x21e2ec4]: pmap_runner_one_arg_t<std::_Bind<std::_Mem_fn<void (connectivity_cluster_t::run_t::*)(peer_address_t const*, int, boost::optional<peer_id_t>, auto_drainer_t::lock_t, bool*, co_semaphore_t*)> (connectivity_cluster_t::run_t*, peer_address_t const*, std::_Placeholder<1>, boost::optional<peer_id_t>, auto_drainer_t::lock_t, bool*, static_semaphore_t*)>, long>::operator()() at pmap.hpp:21
       16 [0x21e2df9]: void pmap<std::_Bind<std::_Mem_fn<void (connectivity_cluster_t::run_t::*)(peer_address_t const*, int, boost::optional<peer_id_t>, auto_drainer_t::lock_t, bool*, co_semaphore_t*)> (connectivity_cluster_t::run_t*, peer_address_t const*, std::_Placeholder<1>, boost::optional<peer_id_t>, auto_drainer_t::lock_t, bool*, static_semaphore_t*)> >(long, long, std::_Bind<std::_Mem_fn<void (connectivity_cluster_t::run_t::*)(peer_address_t const*, int, boost::optional<peer_id_t>, auto_drainer_t::lock_t, bool*, co_semaphore_t*)> (connectivity_cluster_t::run_t*, peer_address_t const*, std::_Placeholder<1>, boost::optional<peer_id_t>, auto_drainer_t::lock_t, bool*, static_semaphore_t*)> const&) at pmap.hpp:47
       17 [0x21d6aba]: void pmap<std::_Bind<std::_Mem_fn<void (connectivity_cluster_t::run_t::*)(peer_address_t const*, int, boost::optional<peer_id_t>, auto_drainer_t::lock_t, bool*, co_semaphore_t*)> (connectivity_cluster_t::run_t*, peer_address_t const*, std::_Placeholder<1>, boost::optional<peer_id_t>, auto_drainer_t::lock_t, bool*, static_semaphore_t*)> >(long, std::_Bind<std::_Mem_fn<void (connectivity_cluster_t::run_t::*)(peer_address_t const*, int, boost::optional<peer_id_t>, auto_drainer_t::lock_t, bool*, co_semaphore_t*)> (connectivity_cluster_t::run_t*, peer_address_t const*, std::_Placeholder<1>, boost::optional<peer_id_t>, auto_drainer_t::lock_t, bool*, static_semaphore_t*)> const&) at pmap.hpp:54
       18 [0x21d2967]: connectivity_cluster_t::run_t::join_blocking(peer_address_t, boost::optional<peer_id_t>, auto_drainer_t::lock_t) at cluster.cc:376
       19 [0x21dda9d]: std::_Mem_fn<void (connectivity_cluster_t::run_t::*)(peer_address_t, boost::optional<peer_id_t>, auto_drainer_t::lock_t)>::operator()(connectivity_cluster_t::run_t*, peer_address_t, boost::optional<peer_id_t>, auto_drainer_t::lock_t) const at functional:550
       20 [0x21e4e19]: _ZNSt5_BindIFSt7_Mem_fnIMN22connectivity_cluster_t5run_tEFv14peer_address_tN5boost8optionalI9peer_id_tEEN14auto_drainer_t6lock_tEEEPS2_S3_MNS4_6detail11none_helperEiS9_EE6__callIvJEJLi0ELi1ELi2ELi3EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0x159 at functional:1145
       21 [0x21e4cb6]: _ZNSt5_BindIFSt7_Mem_fnIMN22connectivity_cluster_t5run_tEFv14peer_address_tN5boost8optionalI9peer_id_tEEN14auto_drainer_t6lock_tEEEPS2_S3_MNS4_6detail11none_helperEiS9_EEclIJEvEET0_DpOT_+0x26 at functional:1203
       22 [0x21e4bdc]: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (connectivity_cluster_t::run_t::*)(peer_address_t, boost::optional<peer_id_t>, auto_drainer_t::lock_t)> (connectivity_cluster_t::run_t*, peer_address_t, int boost::detail::none_helper::*, auto_drainer_t::lock_t)> >::run_action() at callable_action.hpp:32
       23 [0x165cd12]: callable_action_wrapper_t::run() at runtime_utils.cc:48
       24 [0x164ed56]: coro_t::run() at coroutines.cc:205
       25 [0x21d688d]: void coro_t::spawn_now_dangerously<std::_Bind<std::_Mem_fn<void (connectivity_cluster_t::run_t::*)(peer_address_t, boost::optional<peer_id_t>, auto_drainer_t::lock_t)> (connectivity_cluster_t::run_t*, peer_address_t, int boost::detail::none_helper::*, auto_drainer_t::lock_t)> >(std::_Bind<std::_Mem_fn<void (connectivity_cluster_t::run_t::*)(peer_address_t, boost::optional<peer_id_t>, auto_drainer_t::lock_t)> (connectivity_cluster_t::run_t*, peer_address_t, int boost::detail::none_helper::*, auto_drainer_t::lock_t)>&&) at coroutines.hpp:57
       26 [0x21cf615]: connectivity_cluster_t::run_t::join(peer_address_t const&) at cluster.cc:283
       27 [0x17b128c]: initial_joiner_t::main_coro(connectivity_cluster_t::run_t*, auto_drainer_t::lock_t) at initial_join.cc:104
       28 [0x17b27a9]: boost::_mfi::mf2<void, initial_joiner_t, connectivity_cluster_t::run_t*, auto_drainer_t::lock_t>::operator()(initial_joiner_t*, connectivity_cluster_t::run_t*, auto_drainer_t::lock_t) const at mem_fn_template.hpp:280
       29 [0x17b26b9]: void boost::_bi::list3<boost::_bi::value<initial_joiner_t*>, boost::_bi::value<connectivity_cluster_t::run_t*>, boost::_bi::value<auto_drainer_t::lock_t> >::operator()<boost::_mfi::mf2<void, initial_joiner_t, connectivity_cluster_t::run_t*, auto_drainer_t::lock_t>, boost::_bi::list0>(boost::_bi::type<void>, boost::_mfi::mf2<void, initial_joiner_t, connectivity_cluster_t::run_t*, auto_drainer_t::lock_t>&, boost::_bi::list0&, int) at bind.hpp:392
       30 [0x17b260f]: boost::_bi::bind_t<void, boost::_mfi::mf2<void, initial_joiner_t, connectivity_cluster_t::run_t*, auto_drainer_t::lock_t>, boost::_bi::list3<boost::_bi::value<initial_joiner_t*>, boost::_bi::value<connectivity_cluster_t::run_t*>, boost::_bi::value<auto_drainer_t::lock_t> > >::operator()() at bind_template.hpp:20
       31 [0x17b251c]: callable_action_instance_t<boost::_bi::bind_t<void, boost::_mfi::mf2<void, initial_joiner_t, connectivity_cluster_t::run_t*, auto_drainer_t::lock_t>, boost::_bi::list3<boost::_bi::value<initial_joiner_t*>, boost::_bi::value<connectivity_cluster_t::run_t*>, boost::_bi::value<auto_drainer_t::lock_t> > > >::run_action() at callable_action.hpp:32
       32 [0x165cd12]: callable_action_wrapper_t::run() at runtime_utils.cc:48
       33 [0x164ed56]: coro_t::run() at coroutines.cc:205
       34 [0x17b1dfd]: coro_t* coro_t::spawn_sometime<boost::_bi::bind_t<void, boost::_mfi::mf2<void, initial_joiner_t, connectivity_cluster_t::run_t*, auto_drainer_t::lock_t>, boost::_bi::list3<boost::_bi::value<initial_joiner_t*>, boost::_bi::value<connectivity_cluster_t::run_t*>, boost::_bi::value<auto_drainer_t::lock_t> > > >(boost::_bi::bind_t<void, boost::_mfi::mf2<void, initial_joiner_t, connectivity_cluster_t::run_t*, auto_drainer_t::lock_t>, boost::_bi::list3<boost::_bi::value<initial_joiner_t*>, boost::_bi::value<connectivity_cluster_t::run_t*>, boost::_bi::value<auto_drainer_t::lock_t> > >&&) at coroutines.hpp:63
       35 [0x17b0dcd]: initial_joiner_t::initial_joiner_t(connectivity_cluster_t*, connectivity_cluster_t::run_t*, peer_address_set_t const&, int) at initial_join.cc:89
       36 [0x17e6b44]: do_serve(io_backender_t*, bool, base_path_t const&, metadata_file_t*, serve_info_t const&, os_signal_cond_t*) at serve.cc:244
       37 [0x17e977c]: serve(io_backender_t*, base_path_t const&, metadata_file_t*, serve_info_t const&, os_signal_cond_t*) at serve.cc:646
       38 [0x17c1625]: run_rethinkdb_serve(base_path_t const&, serve_info_t*, file_direct_io_mode_t, int, boost::optional<boost::optional<unsigned long> > const&, uuid_u const*, server_config_versioned_t const*, cluster_semilattice_metadata_t const*, directory_lock_t*, bool*) at command_line.cc:836
       39 [0x17c18e2]: run_rethinkdb_porcelain(base_path_t const&, name_string_t const&, std::set<name_string_t, std::less<name_string_t>, std::allocator<name_string_t> > const&, file_direct_io_mode_t, int, boost::optional<boost::optional<unsigned long> > const&, bool, serve_info_t*, directory_lock_t*, bool*) at command_line.cc:865
       40 [0x17d25d2]: _ZNSt5_BindIFPFvRK11base_path_tRK13name_string_tRKSt3setIS3_St4lessIS3_ESaIS3_EE21file_direct_io_mode_tiRKN5boost8optionalINSF_ImEEEEbP12serve_info_tP16directory_lock_tPbES0_S3_SA_SD_iSH_bSL_SN_SO_EE6__callIvJEJLi0ELi1ELi2ELi3ELi4ELi5ELi6ELi7ELi8ELi9EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0x262 at functional:1145
       41 [0x17d2366]: _ZNSt5_BindIFPFvRK11base_path_tRK13name_string_tRKSt3setIS3_St4lessIS3_ESaIS3_EE21file_direct_io_mode_tiRKN5boost8optionalINSF_ImEEEEbP12serve_info_tP16directory_lock_tPbES0_S3_SA_SD_iSH_bSL_SN_SO_EEclIJEvEET0_DpOT_+0x26 at functional:1203
       42 [0x17d08dd]: std::_Function_handler<void (), std::_Bind<void (*(base_path_t, name_string_t, std::set<name_string_t, std::less<name_string_t>, std::allocator<name_string_t> >, file_direct_io_mode_t, int, boost::optional<boost::optional<unsigned long> >, bool, serve_info_t*, directory_lock_t*, bool*))(base_path_t const&, name_string_t const&, std::set<name_string_t, std::less<name_string_t>, std::allocator<name_string_t> > const&, file_direct_io_mode_t, int, boost::optional<boost::optional<unsigned long> > const&, bool, serve_info_t*, directory_lock_t*, bool*)> >::_M_invoke(std::_Any_data const&) at functional:1779
       43 [0x165b7fe]: std::function<void ()>::operator()() const at functional:2160
       44 [0x165b9e1]: starter_t::run_wrapper(std::function<void ()> const&) at runtime.cc:61
       45 [0x165ca6a]: std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)>::operator()(starter_t*, std::function<void ()> const&) const at functional:550
       46 [0x165c9d0]: _ZNSt5_BindIFSt7_Mem_fnIM9starter_tFvRKSt8functionIFvvEEEEPS1_S4_EE6__callIvJEJLi0ELi1EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0x70 at functional:1145
       47 [0x165c956]: _ZNSt5_BindIFSt7_Mem_fnIM9starter_tFvRKSt8functionIFvvEEEEPS1_S4_EEclIJEvEET0_DpOT_+0x26 at functional:1203
       48 [0x165c07d]: std::_Function_handler<void (), std::_Bind<std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)> (starter_t*, std::function<void ()>)> >::_M_invoke(std::_Any_data const&) at functional:1779
       49 [0x165b7fe]: std::function<void ()>::operator()() const at functional:2160
       50 [0x165b749]: callable_action_instance_t<std::function<void ()>&>::run_action() at callable_action.hpp:32
       51 [0x165cd12]: callable_action_wrapper_t::run() at runtime_utils.cc:48
       52 [0x164ed56]: coro_t::run() at coroutines.cc:205
       53 [0x165b4ed]: coro_t* coro_t::spawn_sometime<std::function<void ()>&>(std::function<void ()>&&&) at coroutines.hpp:63
       54 [0x165b495]: starter_t::on_thread_switch() at runtime.cc:57
       55 [0x165db60]: linux_message_hub_t::on_event(int) at message_hub.cc:154
       56 [0x1644d87]: epoll_event_queue_t::run() at epoll.cc:115
       57 [0x16583e1]: linux_thread_pool_t::start_thread(void*) at thread_pool.cc:172
       58 [0x7fa73fff8e9a]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x7e9a) [0x7fa73fff8e9a] at 0x7fa73fff8e9a (/lib/x86_64-linux-gnu/libpthread.so.0)
       59 [0x7fa73fd2638d]: clone+0x6d at 0x7fa73fd2638d (/lib/x86_64-linux-gnu/libc.so.6)

**Comment by danielmewes** [comment]
Moving to `backlog` since we have only seen this a single time and it's not reproducible as far as I know. Please correct me if I'm wrong.
